### PR TITLE
Adds a binary that can generate PGM/YAML from a .pbstream.

### PR DESCRIPTION
**Pull request by SirVer**

### CODE REVIEW

**Comment by SirVer** [comment]
@jihoonl @ensonic Now, that is the binary.

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+DEFINE_string(pbstream_filename, "",
+              "Filename of a pbstream to draw a map from.");
+DEFINE_string(map_filestem, "map", "Stem of the output files.");
+DEFINE_double(resolution, 0.05, "Resolution of a grid cell in the drawn map.");
+
+namespace cartographer_ros {
+namespace {
+
+void FillSubmapState(const ::cartographer::transform::Rigid3d& submap_pose,
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: See https://google-cartographer.readthedocs.io/en/latest/terminology.html: `global_submap_pose`?
Also: `FillSubmapSlice()`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  submap_slice->surface =
+      DrawTexture(pixels.intensity, pixels.alpha, texture_proto.width(),
+                  texture_proto.height(), &submap_slice->cairo_data);
+}
+
+void Run(const std::string& pbstream_filename, const std::string& map_filestem,
+         const double resolution) {
+  ::cartographer::io::ProtoStreamReader reader(pbstream_filename);
+
+  // Skip Pose graph
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: Full sentences end with a period. Similar below. Also typo: `pose`.

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  ::cartographer::mapping::proto::PoseGraph pose_graph;
+  CHECK(reader.ReadProto(&pose_graph));
+
+  // Load submap_slices from pbstream
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: `'submap_slices'`? Period?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  LOG(INFO) << "Loading submap_slices from serialized data.";
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: Remove underscore or add single quotes?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      submap_slices;
+  for (;;) {
+    ::cartographer::mapping::proto::SerializedData proto;
+    if (!reader.ReadProto(&proto)) {
+      break;
+    }
+    if (proto.has_submap()) {
+      const auto& submap = proto.submap();
+
+      ::cartographer::mapping::SubmapId id{submap.submap_id().trajectory_id(),
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: `const`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                                           submap.submap_id().submap_index()};
+      const ::cartographer::transform::Rigid3d submap_pose =
+          ::cartographer::transform::ToRigid3(
+              pose_graph.trajectory(submap.submap_id().trajectory_id())
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: Conciseness: Use `id.trajectory_id` and `id.submap_index`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                  .submap(submap.submap_id().submap_index())
+                  .pose());
+      FillSubmapState(submap_pose, submap, &submap_slices[id]);
+    }
+  }
+  CHECK(reader.eof());
+
+  LOG(INFO) << "Generate combined map image from submap_slices.";
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: `Generating`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      ::cartographer::io::PaintSubmapSlices(submap_slices, resolution);
+
+  ::cartographer::io::StreamFileWriter pgm_writer(map_filestem + ".pgm");
+
+  ::cartographer::io::Image image(std::move(result.surface));
+  WritePgm(image, resolution, &pgm_writer);
+
+  const Eigen::Vector2d origin(
+      -result.origin.x() * resolution,
+      (-image.height() + result.origin.y()) * resolution);
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: `result.origin.y() - image.height()`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc (hunk 2 above)
Comment (SirVer) [reply cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: removed that comment. It did not add to the narrative of the code.

**Comment by SirVer** [comment]
ptal.

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 9]
```diff
@@ -39,21 +39,22 @@ DEFINE_double(resolution, 0.05, "Resolution of a grid cell in the drawn map.");
 namespace cartographer_ros {
 namespace {
 
-void FillSubmapState(const ::cartographer::transform::Rigid3d& submap_pose,
-                     const ::cartographer::mapping::proto::Submap& proto,
-                     ::cartographer::io::SubmapSlice* submap_slice) {
+void FillSubmapSlice(
+    const ::cartographer::transform::Rigid3d& global_submap_pose,
+    const ::cartographer::mapping::proto::Submap& proto,
+    ::cartographer::io::SubmapSlice* submap_slice) {
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: `* const`?

File: cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+void Run(const std::string& pbstream_filename, const std::string& map_filestem,
+         const double resolution) {
+  ::cartographer::io::ProtoStreamReader reader(pbstream_filename);
+
+  ::cartographer::mapping::proto::PoseGraph pose_graph;
+  CHECK(reader.ReadProto(&pose_graph));
+
+  // Load 'submap_slices' from pbstream.
```
Comment (wohe) [inline cartographer_ros/cartographer_ros/pbstream_to_ros_map_main.cc]: Remove?

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 2.

**Comment by jihoonl** [comment]
I just saw this PR. cool good to know thanks!

**Comment by hadiTab** [comment]
Hey guys, I was wondering how can this be used? Thanks.

**Comment by doisyg** [comment]
`rosrun cartographer_ros cartographer_pbstream_to_ros_map -pbstream_filename=thenameofthefile.pbstream'`

**Pull request merged**